app/models/concerns/course/assessment/submission/workflow_event_concern.rb

```diff
@@ -142,6 +145,57 @@ def unsubmit_current_answers(only_programming: false)
     end
   end
 
+  # When a submission is unsubmitted, every current_answer is copied as and flagged as attempting.
+  # The new copied answer is then marked as current_answer which is the answer that can be modified
+  # by users. The old current_answer is unmarked as current_answer and is kept as graded past answer.
+  def recreate_current_answers
+    current_answers.reject(&:attempting?).each do |current_answer|
+      new_answer = current_answer.question.attempt(current_answer.submission, current_answer)
+
+      current_answer.current_answer = false
+      new_answer.current_answer = true
+      current_answer.save!
+      new_answer.save!
+    end
+  end
+
+  def finalise_current_answers
```

Comment:
Would be good to include some docs above to briefly describe the logic within this method, since it's not exactly straightforward.

May be good to also link the PDF (or the original Google Slides) that you've included in the PR description. I think that PDF is really quite clear. But might need some changes to generalise it, i.e. instead of showing current vs proposed, just show the proposed.